File: day05/haskell/main.hs
import Data.List (nub, sort)
import System.Posix (sleep)

splitOn :: (Eq a, Show a) => a -> [a] -> [[a]]
splitOn = splitOn' []

splitOn' :: (Eq a, Show a) => [a] -> a -> [a] -> [[a]]
splitOn' acc _ [] = [reverse acc]
splitOn' acc c (x : xs)
  | c == x = reverse acc : splitOn' [] c xs
  | otherwise = splitOn' (x : acc) c xs

pairs :: [a] -> [(a, a)]
pairs [] = []
pairs [_] = error "uneven length in pairs"
pairs (a : b : rest) = (a, b) : pairs rest

headWithDefault :: a -> [a] -> a
headWithDefault def [] = def
headWithDefault _ (x : _) = x

inRange :: Int -> (Int, Int, Int) -> Bool
inRange n (_, srcStart, len) = n >= srcStart && n < srcStart + len

reverseInRange :: Int -> (Int, Int, Int) -> Bool
reverseInRange n (dstStart, _, len) = n >= dstStart && n < dstStart + len

convert :: Int -> [(Int, Int, Int)] -> Int
convert n ranges = n + (dstStart - srcStart)
  where
    (dstStart, srcStart, _) = headWithDefault (0, 0, 0) $ filter (inRange n) ranges

convertBack :: Int -> [(Int, Int, Int)] -> Int
convertBack n ranges = n - (dstStart - srcStart)
  where
    (dstStart, srcStart, _) = headWithDefault (0, 0, 0) $ filter (reverseInRange n) ranges

parseRange :: String -> (Int, Int, Int)
parseRange s = (srcStart, dstStart, len)
  where
    [srcStart, dstStart, len] = map read (words s)

parseInput :: String -> ([Int], [[(Int, Int, Int)]])
parseInput content = (seeds, ranges)
  where
    contentLines = lines content
    seeds = map read $ tail $ words $ head contentLines
    ranges = map (map parseRange . tail) $ splitOn "" $ drop 2 contentLines

computeDiscontinuities :: (Int, Int, Int) -> [Int]
computeDiscontinuities (_, srcStart, len) = [srcStart - 1, srcStart, srcStart + len, srcStart + len + 1]

partOne :: ([Int], [[(Int, Int, Int)]]) -> IO ()
partOne (seeds, ranges) = print $ minimum $ map (\seed -> foldl convert seed ranges) seeds

partTwo :: ([Int], [[(Int, Int, Int)]]) -> IO ()
partTwo (seeds, ranges) = do
  print $ minimum $ map (\seed -> foldl convert seed ranges) inputs
  where
    discontinuities = nub $ sort $ foldr (\r cum -> map (`convertBack` r) cum ++ concatMap computeDiscontinuities r) [] ranges
    seedRanges = pairs seeds
    inputs = sort $ concatMap (\(start, len) -> start : start + len : filter (\x -> x >= start && x < start + len) discontinuities) seedRanges

main = do
  content <- readFile "../input.txt"
  let input = parseInput content
  partOne input
  partTwo input

File: day05/julia/main.jl
struct Range
    destinationStart::Int64
    sourceStart::Int64
    length::Int64
end

function readInput(filename::String)::Tuple{Vector{Int64},Vector{Vector{Range}}}
    lines = readlines(filename)
    seeds = map(w -> parse(Int64, w), split(lines[1])[2:end])

    # add empty line at the end
    push!(lines, "")

    transforms = []
    acc::Vector{String} = []
    for i in eachindex(lines)[3:end]
        if lines[i] == ""
            push!(transforms, [Range([parse(Int64, w) for w in split(l)]...) for l in acc[2:end]])
            acc = []
        else
            push!(acc, lines[i])
        end
    end

    return (seeds, transforms)
end

function convert(x::Int64, ranges::Vector{Range})::Int64
    for range in ranges
        if range.sourceStart <= x < range.sourceStart + range.length
            return x + range.destinationStart - range.sourceStart
        end
    end
    return x
end

function convertBack(x::Int64, ranges::Vector{Range})::Int64
    for range in ranges
        if range.destinationStart <= x < range.destinationStart + range.length
            return x - (range.destinationStart - range.sourceStart)
        end
    end
    return x
end

function calculateDiscontinuities(ranges::Vector{Range})::Vector{Int64}
    [n for r in ranges for n in [r.sourceStart - 1, r.sourceStart, r.sourceStart + r.length, r.sourceStart + r.length + 1]]
end

function partOne(seeds::Vector{Int64}, transforms::Vector{Vector{Range}})::Int64
    locations = []
    for s in seeds
        for t in transforms
            s = convert(s, t)
        end
        push!(locations, s)
    end

    minimum(locations)
end

function partTwo(seeds::Vector{Int64}, transforms::Vector{Vector{Range}})
    discontinuities = []
    for t in reverse(transforms)
        discontinuities = map(x -> convertBack(x, t), discontinuities)
        append!(discontinuities, calculateDiscontinuities(t))
    end

    discontinuities = sort(unique(discontinuities))

    inputs::Vector{Int64} = []
    for i in 1:2:length(seeds)
        start = seeds[i]
        length = seeds[i+1]

        append!(inputs, filter(x -> start <= x < start + length, discontinuities))
        push!(inputs, start)
        push!(inputs, start + length)
    end

    partOne(inputs, transforms)
end

seeds, transforms = readInput("../input.txt")
println("Part One: ", partOne(seeds, transforms))
println("Part Two: ", partTwo(seeds, transforms))
